### ISSUE TITLE
[OSS] Support seamless state consolidation

### ISSUE THREAD

**Issue by blefaudeux** [facebookresearch/fairscale#116]
## 🚀 Feature
Currently  `state_dict()` being called on a single rank cannot return the full optimizer state, because this rank only has a shard's worth. Checkpointing the full optimizer state requires calling `consolidate_state()` on all rank prior to pulling the state, which is not intuitive and not pytorch compliant.

Taking inspiration from other projects sharding the model or parameter state, we could use the torch RPC framework to store remote references to other ranks' states, which would allow for a seamless `to_here()` consolidation. In that case, we would need to suppose that both RPC and torch distributed are initialized, which is a feature being tackled.

## Motivation
Pytorch compliance
Smaller API footprint 
Ease of use

## Pitch
See the feature description, this is doable and would be a lot more elegant.

## Alternatives
Status quo

## Additional context
Discussed with Pytorch Distributed and MSFT folks

**Comment by blefaudeux** [comment]
cc @min-xu-ai, feel free to grab (or not), any action equally welcome

**Comment by min-xu-ai** [comment]
Ben, Myle just merged two changes for FSDP on state_dict. We used pre/post state_dict hook. You may want to check it out to see if it helps here.

**Comment by blefaudeux** [comment]
> Ben, Myle just merged two changes for FSDP on state_dict. We used pre/post state_dict hook. You may want to check it out to see if it helps here.

FSDP state dict docstring mentions "this needs to be called on all ranks", which points to a fairly different behaviour (but tied to the same limitations, collective communication primitives). Many frameworks checkpoint from a single rank, because they just want one state to handle down the line (-> push to a bucket), and because historically it was enough. This issue was about tracking solutions for that, they exist (remote references for instance). 

Right now for OSS there's the "consolidate_state_dict()" which needs to be called on all ranks (new call which tentatively does what it says), and state_dict() which tried to follow the pytorch convention (ie: ok to call it from one rank only, and returns the state dict). Looks like FSDP fuses both, which is fine but maybe a bit confusing if you don't actually want the state dict spit out on all ranks ? (and not compliant with other state_dict() calls within PyTorch, but that's probably ok) The solution that this issue called for is a bit better I think.

**Comment by min-xu-ai** [comment]
Yeah, good point. I didn't see that. For FSDP, user perhaps need to change their checkpoint code slightly and only keep the results on rank 0.

**Comment by blefaudeux** [comment]
I think it would be worth it having a remote rank invariant "tracker" actually, it's pretty convenient for the user in that the sharding is completely abstracted away, the issue with [remote refs](https://pytorch.org/docs/stable/rpc/rref.html?highlight=remote%20references) initially was that it requires RPC being initialized and that the init was different from c10 (you had to init both), so it was a major inconvenience for some users (turns out, it's complicated to have something reliably work for various environments it seems) which was not really worth it for an arguably minor convenience :) It's supposed to have improved over time (I could look the issue up, it was tracked on PyTorch side), so could be worth it now ? cc @myleott, heads up
